File: 프로그래머스/2/87946. 피로도/피로도.py
def solution(k, dungeons):
    n = len(dungeons)
    visited = [False for _ in range(n)]
    answer = bt(k, dungeons, visited, 0, 0)
    return answer

def bt(health, dungeons, visited, depth, cnt):
    """
    남은 피로도, 던전 배열, 방문 여부, 탐색 깊이, 탐험한 던전 개수
    """
    
    # 종료 조건 - 모든 던전 탐색했을 때
    if depth == len(dungeons):
        return cnt
    
    max_cnt = cnt  # 현재 cnt를 최댓값으로 초기화
    
    for i in range(len(dungeons)):
        if visited[i]:
            continue
        
        visited[i] = True
        
        health_cutline = dungeons[i][0]
        health_cost = dungeons[i][1]
        
        # 현재 던전 탐험 가능한 체력이면 탐험함
        if health >= health_cutline:
            result = bt(health - health_cost, dungeons, visited, depth + 1, cnt + 1)
            max_cnt = max(max_cnt, result)        
            # 현재 던전 탐험하지 않는 경우
            result = bt(health, dungeons, visited, depth + 1, cnt)
            max_cnt = max(max_cnt, result)
        
        visited[i] = False
    
    return max_cnt

File: 프로그래머스/2/87946. 피로도/피로도.java
class Solution {
    static int answer = -1;
    public int solution(int k, int[][] dungeons) {
        boolean[] visited = new boolean[dungeons.length];
        bt(0, 0, k, dungeons, visited);
        return answer;
    }
    
    // 탐색 깊이, 방문한 던전수, 피로도, 던전정보배열, 방문기록
    public void bt(int depth, int checkedDungeon, int health, int[][] dungeons, boolean[] visited){
        if(depth == dungeons.length){
            answer = Math.max(answer, checkedDungeon);
            return;
        }
        
        for(int i = 0; i < dungeons.length; i++){
            if(visited[i]) continue;
            
            visited[i] = true;
            if(health >= dungeons[i][0]){                
                bt(depth + 1, checkedDungeon + 1, health - dungeons[i][1], dungeons, visited);                
            } else {
                bt(depth + 1, checkedDungeon, health, dungeons, visited);
            }            
            visited[i] = false;
        }
    }
}